### PR TITLE
Split BinaryProvider into read vs write

### PR DESCRIPTION
**Pull request by ras0219-msft**
This is a rewrite of the Binary Caching infrastructure to improve code sharing and reduce coupling.

This will make #908 more tractable because there will be significantly fewer lines to audit.

### CODE REVIEW

**Comment by autoantwort** [comment]
In the past a provider was marked `available` if the provider pushed a specific package. This was removed [here](https://github.com/microsoft/vcpkg-tool/pull/996/files#diff-045a3245d6bccf63618bd35376d0a973af7805c6e232bbbf222ca326b60b7000L1663) (expand diff required to see the line), resp. in #908 because the call was not thread safe and there is currently no situation were this is needed. It also could be kind of ambiguous because a provider can have multiple destinations and we don't know which destination has a package available. 

With #802 the "`push_success` -> `mark_restored` flow" will become relevant for the first time because we build packages, will remove them and then maybe need them again. If we don't want to rebuild the package although it is in the cache, we have to mark it as available or remove the unavailable status. In #911 there is now one provider for every destination so that we don't lost the connection between the push and pull providers and are able to mark providers as available for a specific package.

**Comment by ras0219-msft** [comment]
It's a really good point that we need to invalidate the precheck status of any packages that were pushed -- otherwise we would refuse to try to download them.

However, we SHOULD be running `.precheck()` on every plan -- which will then correctly re-populate status information. I don't think coupling readers to writers really does what we want, because nothing says that pushing to one reader can't enable pulling from another. I think the correct fix (which you observed) is to completely invalidate the status upon push. Then, when we begin to execute a new plan, we run a `.precheck()` which will reconsider all read providers (in performance order).

**Comment by autoantwort** [comment]
> However, we SHOULD be running `.precheck()` on every plan

This is currently already the case :)

> I don't think coupling readers to writers really does what we want, because nothing says that pushing to one reader can't enable pulling from another.

That is true (but I think this only happens rarely), but at least for the destinations we pushed a package to, we know we can read it back from there. Invalidating also works and is needed for providers to which we don't have pushed a package, but for the other providers that means unnecessary work (but the overhead is probably negligible).

File: src/vcpkg/binarycaching.cpp [hunk 1]
```diff
@@ -2310,137 +1856,393 @@ namespace
...
-            }
-            else if (segments[0].second == "clear")
-            {
-                if (segments.size() >= 2)
+                else if (cache_result[i] == CacheAvailability::unavailable)
                 {
-                    return add_error(msg::format(msgAssetCacheProviderAcceptsNoArguments, msg::value = "clear"),
-                                     segments[1].first);
+                    m_status[*action_ptrs[i]->package_abi().get()].mark_unavailable(provider.get());
                 }
```
Comment (BillyONeal) [inline src/vcpkg/binarycaching.cpp:2077]: Suggestion:

```
                auto&& this_status = m_status[*action_ptrs[i]->package_abi().get()];
                if (cache_result[i] == CacheAvailability::available)
                {
                    this_status.mark_available(provider.get());
                }
                else if (cache_result[i] == CacheAvailability::unavailable)
                {
                    this_status.mark_unavailable(provider.get());
                }
```

File: src/vcpkg/binarycaching.cpp [hunk 2]
```diff
@@ -2310,137 +1856,393 @@ namespace
...
-            script = nullopt;
+            auto it = m_status.find(*abi);
+            if (it != m_status.end()) return it->second.is_restored();
         }
-    };
+        return false;
+    }
 
-    struct AssetSourcesParser : ConfigSegmentsParser
+    std::vector<CacheAvailability> ReadOnlyBinaryCache::precheck(View<InstallPlanAction> actions)
```
Comment (BillyONeal) [inline src/vcpkg/binarycaching.cpp:2040]: I think it's now possible to write tests for this, and that's probably worth it now that this function is doing complicated stuff to map between universes.

File: src/vcpkg/binarycaching.cpp [hunk 3]
```diff
@@ -2310,137 +1856,393 @@ namespace
...
+
+    void BinaryCache::push_success(const InstallPlanAction& action)
+    {
+        if (action.package_abi().has_value())
+        {
+            bool restored = m_status[*action.package_abi().get()].is_restored();
+            // Purge all status information on push_success (cache invalidation)
+            // - push_success may delete packages/ (invalidate restore)
+            // - push_success may make the package available from providers (invalidate unavailable)
+            m_status.erase(*action.package_abi().get());
```
Comment (BillyONeal) [inline src/vcpkg/binarycaching.cpp:2111]:
```suggestion
        if (auto package_abi = action.package_abi.get())
        {
            bool restored = m_status[*package_abi].is_restored();
            // Purge all status information on push_success (cache invalidation)
            // - push_success may delete packages/ (invalidate restore)
            // - push_success may make the package available from providers (invalidate unavailable)
            m_status.erase(*package_abi);
```

**Review by BillyONeal — COMMENTED** [review]
Most of the CR comments are nits. A few non-nits:

* I'm weakly opposed to RemoveFilesystem.
* The comments describing the interface contracts.
* Localizing vendor().

File: src/vcpkg/binarycaching.cpp [hunk 4]
```diff
@@ -206,399 +204,361 @@ namespace
...
+            const auto& zip_path = *request.zip_path.get();
+            size_t count_stored = 0;
+            for (auto&& templ : m_urls)
+            {
+                auto url = templ.instantiate_variables(request);
+                auto maybe_success = put_file(m_fs, url, m_secrets, templ.headers, zip_path);
+                if (maybe_success)
+                    count_stored++;
+                else
+                    msg_sink.println(Color::warning, maybe_success.error());
```
[the remark this answers is not in the document]
Comment (ras0219-msft) [reply src/vcpkg/binarycaching.cpp:411]: `maybe_success` is an `ExpectedT` in this case, and `maybe_success.error()` is used in the `else`. I don't think the suggestion compiles.

File: src/vcpkg/binarycaching.cpp [hunk 5]
```diff
@@ -206,399 +204,361 @@ namespace
...
+        }
 
-        size_t push_success(const BinaryProviderPushRequest&, MessageSink&) override { return 0; }
+        bool needs_nuspec_data() const override { return false; }
+        bool needs_zip_file() const override { return true; }
 
-        void prefetch(View<InstallPlanAction> actions, View<CacheStatus*> cache_status) const override
+    private:
+        const Filesystem& m_fs;
+        std::vector<UrlTemplate> m_urls;
```
[the remark this answers is not in the document]
Comment (ras0219-msft) [reply src/vcpkg/binarycaching.cpp:421]: I generally left the "write" side combined, since there was not a complex state machine to implement between sources.

File: include/vcpkg/archives.h [hunk 6]
```diff
@@ -27,14 +30,23 @@ namespace vcpkg
         Filesystem& fs, const ToolCache& tools, MessageSink& status_sink, const Path& archive, const Path& to_path);
 #endif
 
-    // Compress the source directory into the destination file.
-    ExpectedL<Unit> compress_directory_to_zip(
-        Filesystem& fs, const ToolCache& tools, MessageSink& status_sink, const Path& source, const Path& destination);
+    struct ZipTool
+    {
+        // On Windows, acquires 7zip in the constructor
```
[the remark this answers is not in the document]
Comment (ras0219-msft) [reply include/vcpkg/archives.h]: I believe this is addressed.

File: include/vcpkg/base/files.h [hunk 7]
```diff
@@ -107,10 +107,23 @@ namespace vcpkg
         virtual ExpectedL<std::vector<std::string>> read_lines(const Path& file_path) const = 0;
 
     protected:
-        ~ILineReader();
+        ~ILineReader() = default;
```
[the remark this answers is not in the document]
Comment (ras0219-msft) [reply include/vcpkg/base/files.h]: I believe this is addressed.

File: include/vcpkg/base/files.h [hunk 8]
```diff
@@ -165,13 +178,6 @@ namespace vcpkg
...
                                     std::error_code& ec) = 0;
 
-        virtual bool remove(const Path& target, std::error_code& ec) = 0;
-        bool remove(const Path& target, LineInfo li);
-
-        virtual void remove_all(const Path& base, std::error_code& ec, Path& failure_point) = 0;
-        void remove_all(const Path& base, std::error_code& ec);
-        void remove_all(const Path& base, LineInfo li);
-
         void remove_all_inside(const Path& base, std::error_code& ec, Path& failure_point);
```
[the remark this answers is not in the document]
Comment (ras0219-msft) [reply include/vcpkg/base/files.h]: I believe this is addressed.

File: src/vcpkg/binarycaching.cpp [hunk 9]
```diff
@@ -2310,137 +1856,393 @@ namespace
...
+            };
+            Optional<NugetBaseBinaryProvider> lazy_nuget_base;
+            auto nuget_base = [&]() -> const NugetBaseBinaryProvider& {
+                if (!lazy_nuget_base)
+                    lazy_nuget_base.emplace(
+                        fs, NuGetTool(tools, stdout_sink, s), paths.packages(), buildtrees, s.nuget_prefix);
+                return *lazy_nuget_base.get();
+            };
+
+            for (auto&& dir : s.archives_to_read)
```
[the remark this answers is not in the document]
Comment (ras0219-msft) [reply src/vcpkg/binarycaching.cpp]: I believe this is addressed.

File: src/vcpkg/binarycaching.cpp (hunk 1 above)
Comment (ras0219-msft) [reply src/vcpkg/binarycaching.cpp]: I believe this is addressed.

File: include/vcpkg/binarycaching.h [hunk 10]
```diff
@@ -135,48 +152,68 @@ namespace vcpkg
...
         std::unordered_map<std::string, CacheStatus> m_status;
-        std::vector<std::unique_ptr<IBinaryProvider>> m_providers;
-        bool needs_nuspec_data = false;
-        Filesystem& filesystem;
+    };
+
+    struct BinaryCache : ReadOnlyBinaryCache
+    {
+        BinaryCache(RemoveFilesystem& fs);
+        BinaryCache(const VcpkgCmdArguments& args, const VcpkgPaths& paths, const LineInfo& info);
```
[the remark this answers is not in the document]
Comment (ras0219-msft) [reply include/vcpkg/binarycaching.h]: I believe this is addressed.

File: src/vcpkg/commands.install.cpp [hunk 11]
```diff
@@ -1279,8 +1271,9 @@ namespace vcpkg
 
         track_install_plan(action_plan);
 
-        BinaryCache binary_cache = only_downloads ? BinaryCache(paths.get_filesystem()) : BinaryCache(args, paths);
-        binary_cache.prefetch(action_plan.install_actions);
+        BinaryCache binary_cache =
```
[the remark this answers is not in the document]
Comment (ras0219-msft) [reply src/vcpkg/commands.install.cpp]: I believe this is addressed.

File: include/vcpkg/binarycaching.h [hunk 12]
```diff
@@ -22,83 +24,98 @@ namespace vcpkg
...
+
+        /// Checks whether the `actions` are present in the cache, without restoring them.
+        ///
+        /// Used by CI to determine missing packages. For each `i`, out_status[i] should be set to
+        /// CacheAvailability::available or CacheAvailability::unavailable
+        ///
+        /// Prerequisites: actions[i].package_abi(), out_status.size() == actions.size()
+        virtual void precheck(View<const InstallPlanAction*> actions, Span<CacheAvailability> out_status) const = 0;
+
+        /// <returns>A user-visible vendor string identifying this provider</returns>
```
[the remark this answers is not in the document]
Comment (ras0219-msft) [reply include/vcpkg/binarycaching.h]: I believe this is addressed.

File: src/vcpkg/binarycaching.cpp [hunk 13]
```diff
@@ -206,399 +204,361 @@ namespace
...
+    struct FilesWriteBinaryProvider : IWriteBinaryProvider
+    {
+        FilesWriteBinaryProvider(Filesystem& fs, std::vector<Path>&& dirs) : m_fs(fs), m_dirs(std::move(dirs)) { }
 
-        void prefetch(View<InstallPlanAction> actions, View<CacheStatus*> cache_status) const override
+        size_t push_success(const BinaryPackageWriteInfo& request, MessageSink& msg_sink) override
         {
-            std::vector<size_t> to_try_restore_idxs;
-            std::vector<const InstallPlanAction*> to_try_restore;
+            if (!request.zip_path) return 0;
```
[the remark this answers is not in the document]
Comment (ras0219-msft) [reply src/vcpkg/binarycaching.cpp]: I believe this is addressed.

File: src/vcpkg/binarycaching.cpp [hunk 14]
```diff
@@ -206,399 +204,361 @@ namespace
...
+            const auto& zip_path = request.zip_path.value_or_exit(VCPKG_LINE_INFO);
+            const auto archive_subpath = files_archive_subpath(request.package_abi);
 
-            for (const auto& archives_root_dir : m_read_dirs)
+            int count_stored = 0;
```
[the remark this answers is not in the document]
Comment (ras0219-msft) [reply src/vcpkg/binarycaching.cpp]: I believe this is addressed.

File: src/vcpkg/binarycaching.cpp [hunk 15]
```diff
@@ -206,399 +204,361 @@ namespace
...
+                m_fs.copy_file(zip_path, archive_path, CopyOptions::overwrite_existing, ec);
+                if (ec)
                 {
-                    auto idx_cache_status = cache_status[idx];
-                    if (idx_cache_status && idx_cache_status->should_attempt_restore(this))
-                    {
-                        to_try_restore_idxs.push_back(idx);
-                        to_try_restore.push_back(&actions[idx]);
-                    }
+                    ec.clear();
```
[the remark this answers is not in the document]
Comment (ras0219-msft) [reply src/vcpkg/binarycaching.cpp]: I believe this is addressed.

File: src/vcpkg/binarycaching.cpp [hunk 16]
```diff
@@ -206,399 +204,361 @@ namespace
...
+        ZipResource(Path&& p, ToRemove t) : path(std::move(p)), to_remove(t) { }
+
+        Path path;
+        ToRemove to_remove;
+    };
+
+    // This middleware class contains logic for BinaryProviders that operate on zip files.
+    // Derived classes must implement:
+    // - acquire_zips()
+    // - post_decompress()
```
[the remark this answers is not in the document]
Comment (ras0219-msft) [reply src/vcpkg/binarycaching.cpp]: I believe this is addressed.

File: src/vcpkg/binarycaching.cpp [hunk 17]
```diff
@@ -206,399 +204,361 @@ namespace
...
+    private:
+        Filesystem& m_fs;
+        std::vector<Path> m_dirs;
+    };
+
+    enum class ToRemove
+    {
+        nothing,
+        remove_always,
+        remove_on_fail,
```
[the remark this answers is not in the document]
Comment (ras0219-msft) [reply src/vcpkg/binarycaching.cpp]: I believe this is addressed.

File: src/vcpkg/binarycaching.cpp [hunk 18]
```diff
@@ -206,399 +204,361 @@ namespace
...
-            return RestoreResult::unavailable;
         }
 
-        size_t push_success(const BinaryProviderPushRequest& request, MessageSink& msg_sink) override
-        {
-            if (m_write_dirs.empty() && m_put_url_templates.empty())
-            {
-                return 0;
-            }
+        // Download/find the zip files corresponding with a set of actions
```
[the remark this answers is not in the document]
Comment (ras0219-msft) [reply src/vcpkg/binarycaching.cpp]: I believe this is addressed.

File: src/vcpkg/binarycaching.cpp [hunk 19]
```diff
@@ -206,399 +204,361 @@ namespace
...
-                {
-                    cache_status[idx]->mark_unavailable(this);
-                }
+                out_status[i] = codes[i] == 200 ? CacheAvailability::available : CacheAvailability::unavailable;
             }
         }
 
-        const VcpkgPaths& paths;
-        std::vector<UrlTemplate> m_url_templates;
+        StringView vendor() const override { return "HTTP servers"; }
```
[the remark this answers is not in the document]
Comment (ras0219-msft) [reply src/vcpkg/binarycaching.cpp]: I believe this is addressed.

File: src/vcpkg/binarycaching.cpp [hunk 20]
```diff
@@ -206,399 +204,361 @@ namespace
...
-                    clean_prepare_dir(fs, action.package_dir.value_or_exit(VCPKG_LINE_INFO));
-                    auto uri = url_template.instantiate_variables(BinaryPackageInformation{action, ""});
-                    url_paths.emplace_back(std::move(uri), make_temp_archive_path(paths.buildtrees(), action.spec));
-                    url_indices.push_back(idx);
-                }
-
-                if (url_paths.empty()) break;
+                auto&& action = *actions[idx];
+                auto uri = m_url_template.instantiate_variables(BinaryPackageReadInfo{action});
+                url_paths.emplace_back(std::move(uri), make_temp_archive_path(m_buildtrees, action.spec));
```
[the remark this answers is not in the document]
Comment (ras0219-msft) [reply src/vcpkg/binarycaching.cpp]: I believe this is addressed.

File: src/vcpkg/binarycaching.cpp [hunk 21]
```diff
@@ -2310,137 +1856,393 @@ namespace
...
+
+            ret.nuget_prefix = s.nuget_prefix;
+
+            Optional<ZipTool> lazy_zip_tool;
+            auto zip_tool = [&]() -> const ZipTool& {
+                if (!lazy_zip_tool) lazy_zip_tool.emplace(fs, tools, stdout_sink);
+                return *lazy_zip_tool.get();
+            };
+
+            std::shared_ptr<const GcsStorageTool> lazy_gcs_tool;
```
[the remark this answers is not in the document]
Comment (ras0219-msft) [reply src/vcpkg/binarycaching.cpp]: I've reworked this section to avoid the lambdas; please take a look.

File: src/vcpkg/binarycaching.cpp (hunk 3 above)
Comment (ras0219-msft) [reply src/vcpkg/binarycaching.cpp]: I believe this has been addressed.

File: src/vcpkg/binarycaching.cpp [hunk 22]
```diff
@@ -2310,137 +1856,393 @@ namespace
...
-
-                p.append("<SHA>");
-                if (segments.size() > 2 && !segments[2].second.empty())
+                if (m_needs_zip_file)
                 {
-                    if (!Strings::starts_with(segments[2].second, "?"))
+                    Path zip_path = request.package_dir + ".zip";
+                    auto compress_result = m_zip_tool.value_or_exit(VCPKG_LINE_INFO)
+                                               .compress_directory_to_zip(request.package_dir, zip_path);
+                    if (!compress_result)
```
[the remark this answers is not in the document]
Comment (ras0219-msft) [reply src/vcpkg/binarycaching.cpp]: I believe this has been addressed.

File: src/vcpkg/binarycaching.cpp [hunk 23]
```diff
@@ -1011,196 +836,150 @@ namespace
...
-            : paths(paths), m_read_prefixes(std::move(read_prefixes)), m_write_prefixes(std::move(write_prefixes))
+        ObjectStorageProvider(ZipTool zip,
+                              Filesystem& fs,
+                              const Path& buildtrees,
+                              std::string&& prefix,
+                              std::shared_ptr<const IObjectStorageTool> tool)
+            : ZipReadBinaryProvider(std::move(zip), fs)
+            , m_buildtrees(buildtrees)
+            , m_prefix(std::move(prefix))
+            , m_tool(std::move(tool))
```
[the remark this answers is not in the document]
Comment (ras0219-msft) [reply src/vcpkg/binarycaching.cpp]: I believe this has been addressed.

File: src/vcpkg/binarycaching.cpp [hunk 24]
```diff
@@ -663,246 +623,145 @@ namespace
...
-                    .string_arg("-ForceEnglishOutput");
-                if (!m_interactive)
-                {
-                    cmdline.string_arg("-NonInteractive");
-                }
-                if (!m_use_nuget_cache)
-                {
-                    cmdline.string_arg("-DirectDownload").string_arg("-NoCache");
-                }
+                Checks::unreachable(
```
[the remark this answers is not in the document]
Comment (ras0219-msft) [reply src/vcpkg/binarycaching.cpp]: I believe this has been addressed.

File: src/vcpkg/commands.set-installed.cpp [hunk 25]
```diff
@@ -118,10 +118,14 @@ namespace vcpkg::Commands::SetInstalled
         if (auto p_pkgsconfig = maybe_pkgsconfig.get())
         {
             auto pkgsconfig_path = paths.original_cwd / *p_pkgsconfig;
-            auto pkgsconfig_contents = generate_nuget_packages_config(action_plan);
+            auto pkgsconfig_contents = generate_nuget_packages_config(action_plan, args.nuget_id_prefix.value_or(""));
             fs.write_contents(pkgsconfig_path, pkgsconfig_contents, VCPKG_LINE_INFO);
             msg::println(msgWroteNuGetPkgConfInfo, msg::path = pkgsconfig_path);
         }
+        else if (dry_run != DryRun::Yes)
```
[the remark this answers is not in the document]
Comment (ras0219-msft) [reply src/vcpkg/commands.set-installed.cpp:125]: This was due to a bad merge on my part -- the ABIs are already calculated above on L100.

File: .vscode/settings.json [hunk 26]
```diff
@@ -31,5 +31,73 @@
     "source.organizeImports": true
   },
   "eslint.format.enable": true,
-  "files.eol": "\n"
+  "files.eol": "\n",
+  "files.associations": {
+    "any": "cpp",
```
Comment (BillyONeal) [inline .vscode/settings.json:36]: Hmmm was this edit intended? .any isn't usually a C++ file

File: src/vcpkg/base/files.cpp [hunk 27]
```diff
@@ -293,9 +293,9 @@ namespace
 #if defined(_WIN32)
         const auto ads = std::find(filename, last, ':'); // strip alternate data streams in intra-filename decomposition
         const auto extension = find_extension(filename, ads);
-#else  // ^^^ _WIN32 / !_WIN32 vvv
+#else                                                    // ^^^ _WIN32 / !_WIN32 vvv
```
Comment (BillyONeal) [inline src/vcpkg/base/files.cpp:296]: You need to explicitly install Clang 15.0.7 to be consistent with the formatters in CI.

File: src/vcpkg/binarycaching.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (BillyONeal) [reply src/vcpkg/binarycaching.cpp:407]: Hmmm no idea why I thought that was a bool :)

File: src/vcpkg/binarycaching.cpp (hunk 21 above)
[the remark this answers is not in the document]
Comment (BillyONeal) [reply src/vcpkg/binarycaching.cpp]: To be clear I was more trying to avoid the repetition of the 'algorithm' rather than lambdas (although I do have a preference to avoid lambdas where possible because debuggers hate them).

I think this one has been addressed with your rework 👍

File: src/vcpkg/binarycaching.cpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (BillyONeal) [reply src/vcpkg/binarycaching.cpp]: Sorry, I know you like avoiding the {}s.

Maybe we could argue that they aren't necessary because we force clang-format? To be clear, I used to be anti {}s until https://www.synopsys.com/blogs/software-security/understanding-apple-goto-fail-vulnerability-2/

**Review by BillyONeal — APPROVED** [review]
LGTM as long as the files.associations thing that looks unintended (?) is fixed.

File: src/vcpkg/binarycaching.cpp [hunk 28]
```diff
@@ -663,246 +623,142 @@ namespace
...
+            auto packed_result = m_cmd.pack(msg_sink, nuspec_path, m_buildtrees);
+            m_fs.remove(nuspec_path, IgnoreErrors{});
+            if (!packed_result)
+            {
+                msg_sink.println(Color::error, msgPackingVendorFailed, msg::vendor = "NuGet");
+                return 0;
             }
 
-            for (auto&& cfg : m_read_configs)
+            int count_stored = 0;
```
Comment (BillyONeal) [inline src/vcpkg/binarycaching.cpp:755]:
```suggestion
            size_t count_stored = 0;
```

File: .vscode/settings.json (hunk 26 above)
Comment (ras0219-msft) [reply .vscode/settings.json:36]: Yes, codespaces keeps making these changes automatically and I can't figure out how to suppress it :(

**Comment by BillyONeal** [comment]
I'm sorry, I tried https://github.com/microsoft/vcpkg-tool/assets/1544943/3b3f5fd8-254a-4991-ac1f-ab6bdca97eda

File: src/vcpkg/binarycaching.cpp (hunk 28 above)
Comment (ras0219-msft) [reply src/vcpkg/binarycaching.cpp]: I believe this has been addressed.

File: include/vcpkg/archives.h [hunk 29]
```diff
@@ -27,14 +30,23 @@ namespace vcpkg
...
+        ZipTool(RemoveFilesystem& fs, const ToolCache& tools, MessageSink& status_sink);
 
-    Command decompress_zip_archive_cmd(const ToolCache& tools,
-                                       MessageSink& status_sink,
-                                       const Path& dst,
-                                       const Path& archive_path);
+    private:
+        RemoveFilesystem* fs;
```
[the remark this answers is not in the document]
Comment (ras0219-msft) [reply include/vcpkg/archives.h]: I believe this has been addressed.

**Comment by ras0219-msft** [comment]
Items left unaddressed:
- `ReadOnlyBinaryCache::precheck` should be more testable now -- we should add more tests
- Elapsed times are suboptimally localized
